src/main/scala/io/iohk/ethereum/jsonrpc/server/http/JsonRpcHttpServer.scala

```diff
@@ -45,24 +51,57 @@ trait JsonRpcHttpServer extends Json4sSupport {
       }
       .result()
 
-  val route: Route = cors(corsSettings) {
-    handleRejections(myRejectionHandler) {
-      (path("healthcheck") & pathEndOrSingleSlash & get) {
-        handleHealthcheck()
+  val myExceptionHandler: ExceptionHandler =
+    ExceptionHandler {
+      case EntityStreamSizeException(limit, actualSize) =>
+        val msg = s"Entity size ${actualSize.getOrElse("(?)")} exceeds limit of ${limit}"
+        val error = JsonRpcErrors.InvalidRequest.copy(data = Some(JString(msg)))
+        complete((StatusCodes.BadRequest, JsonRpcResponse("2.0", None, Some(error), JInt(0))))
+    }
+
+  val mainRoute: Route = {
+    (path("healthcheck") & pathEndOrSingleSlash & get) {
+      handleHealthcheck()
+    } ~
+      (path("buildinfo") & pathEndOrSingleSlash & get) {
+        handleBuildInfo()
       } ~
-        (path("buildinfo") & pathEndOrSingleSlash & get) {
-          handleBuildInfo()
-        } ~
-        (pathEndOrSingleSlash & post) {
-          entity(as[JsonRpcRequest]) { request =>
-            handleRequest(request)
-          } ~ entity(as[Seq[JsonRpcRequest]]) { request =>
-            handleBatchRequest(request)
-          }
+      (pathEndOrSingleSlash & post) {
+        entity(as[JsonRpcRequest]) { request =>
+          handleRequest(request)
+        } ~ entity(as[Seq[JsonRpcRequest]]) { request =>
+          handleBatchRequest(request)
         }
-    }
+      }
   }
 
+  val stricEntityDuration = FiniteDuration(2 * 150, TimeUnit.MILLISECONDS)
+
+  val limitsRoute: Route =
+    withSizeLimit(maxContentLength) {
+      toStrictEntity(stricEntityDuration) {
+        extractRequestEntity {
+          case HttpEntity.Strict(_, data) ⇒
+            metrics.RequestSizeDistribution.record(data.size)
+
+            mainRoute
+
+          case _ ⇒
+            assert(false)
+            mainRoute
+        }
+      }
+    }
+
+  val route: Route =
+    cors(corsSettings) {
+      handleRejections(myRejectionHandler) {
+        handleExceptions(myExceptionHandler) {
+          limitsRoute
```

Comment:
This probably comes with a cost for some edgy edge cases... could we do `if (maxContentLength > 0) limitsRoute else mainRoute` ?